src/charm.py

```diff
@@ -10,41 +10,68 @@
 import os
 from charms.operator_libs_linux.v2 import snap  # type: ignore
 from ops.model import ActiveStatus, MaintenanceStatus
-from snap_management import SnapSpecError, SnapInstallError, SnapServiceError, install_snap, node_exporter_snap_name, opentelemetry_collector_snap_name
+from snap_management import (
+    SnapSpecError,
+    SnapInstallError,
+    SnapServiceError,
+    install_snap,
+    node_exporter_snap_name,
+    opentelemetry_collector_snap_name,
+)
+from singleton_snap import SingletonSnapManager
 
 # Log messages can be retrieved using juju debug-log
 logger = logging.getLogger(__name__)
 VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
 
 SNAPS = [opentelemetry_collector_snap_name, node_exporter_snap_name]
 
+
 class OpentelemetryCollectorOperatorCharm(ops.CharmBase):
     """Charm the service."""
 
     def __init__(self, framework: ops.Framework):
         super().__init__(framework)
         self._reconcile()
 
-
     def _reconcile(self):
         self._install()
-        self._remove()
+        self._stop()
         self.unit.status = ActiveStatus()
 
     def _install(self) -> None:
         if self.hook != "install":
             return
 
+        manager = SingletonSnapManager(self.unit.name)
         for snap_package in SNAPS:
-            self._install_snap(snap_package)
-            self._start_snap(snap_package)
-
-    def _remove(self) -> None:
-        if self.hook != "remove":
+            manager.register(snap_package)
+            with manager.snap_operation(snap_package):
+                self._install_snap(snap_package)
+                self._start_snap(snap_package)
+
+            with manager.config_operation(snap_package):
+                # Merge configurations under a directory into one,
+                # and write it to the default otelcol config file.
+                # This is a placeholder for actual configuration merging logic.
+                # For example:
+                #
+                # content = merge_config()
```

Comment:
Silly thought: Would it be useful if the `registration_file_path` we use per unit actually contains the config itself?
Because we need some prescribed location for the individual config anyway, and since we already have such a prescribed path, then perhaps we could make use of it for the per-unit config?

Reply:
I think it is a good idea, we need a place to store the individual config anyway

Reply:
@sed-i the `registration_file_path` contains the snap revision:

```python
    def register(self, snap_name: str, revision: int) -> None:
        """Register current unit as using the specified snap and revision.

        Args:
            snap_name: Name of the snap.
            revision: Optional revision to put in the lock file. Defaults to an empty string.

        Raises:
            OSError: if there is an I/O related error creating the lock file.
        """
        lock_path = self._get_registration_file_path(snap_name)
        with self._lock_directory():
            with open(lock_path, "w") as f:
                f.write(str(revision))
```

Where do we put the config?